libvisual/libvisual/lv_ringbuffer.c

```diff
@@ -416,8 +426,13 @@ int visual_ringbuffer_get_data_from_end (VisRingBuffer *ringbuffer, VisBuffer *d
 	int totalsize = visual_ringbuffer_get_size (ringbuffer);
 	int offset = totalsize - nbytes;
 
-	if ((nbytes / totalsize) > 0)
+	if ((totalsize != 0) && (nbytes / totalsize) > 0) {
```

Comment:
Since totalsize is positive (when not zero), we can simplify the condition
```suggestion
	if (totalsize != 0 && nbytes > 0) {
```

Reply:
@kaixiong could it be you meant `nbytes >= totalsize` rather than `nbytes > 0`? What is the idea behind `nbytes > 0`?

Reply:
@hartwork, I'm applying a bit of mathematics here. `nbytes / totalsize` is positive if and only if `nbytes` is positive.

Let $t$ be `totalsize`, $n$ be `nbytes`.

For $\frac{n}{t} > 0$ to hold, $n$ and $t$ must have the same sign. Since $t > 0$ , it follows that $n > 0$.

Alternatively, you can manipulate the inequality like this:
$\frac{n}{t} > 0 \implies n > 0 \times t \implies n > 0$
(Multiply both sides by $t$, the > condition is kept because $t > 0$)

Reply:
@kaixiong I see implications here rather than _equality_. `(nbytes / totalsize) > 0` in intergers is equal to `(nbytes / totalsize) >= 1` is equal to `nbytes >= totalsize`. `nbytes > 0` requires less from `nbytes`, it's not equivalent and I'm unsure why that's a beneficial change here. What do you think?

Reply:
@hartwork, I'm only working from the given premise of `(nbytes / totalsize) > 0`.

Your condition is correct too, but more complicated. The reason it's also correct is that we have _integer divisions_.

Reply:
@kaixiong I don't understand. Let's look at an example: for `nbytes = 3, totalsize = 5` we get different results after the suggested patch:

- Before: `(nbytes / totalsize) > 0` yields `false` (because of integer divison)
- After: `nbytes > 0` yields `true`

What am I missing?

Reply:
@hartwork oh wait, my version is wrong because of integer division. You're right, my bad :sweat:

Reply:
@kaixiong no problem. Leave that part as-is then or `nbytes >= totalsize`? (I have a feeling that case `offset == totalsize` may need a closer look as well, but maybe not right now.)

Reply:
@hartwork, yeah let's use `nbytes >= totalsize`. I also agree that it needs a closer look. In fact, I think the whole module could use a closer look.

Anyway, I got a bit too excited to have found an excuse to use LaTeX on Github..

Reply:
@kaixiong I noticed now that `nbytes > totalsize` (rather than `nbytes >= totalsize`) will save us for free in the `totalsize == nbytes` case because `offset := 0` is healthy but `offset := totalsize` may be not…